### PR TITLE
Fix buffered channel bug

### PR DESCRIPTION
**Pull request by yycptt**
Check if there's any blocked sends when getting values from buffer.

### CODE REVIEW

**Comment by yycptt** [comment]
Close the PR for now for adding a unit test.

File: internal/internal_workflow.go [hunk 1]
```diff
@@ -584,6 +585,17 @@ func (c *channelImpl) receiveAsyncImpl(callback *receiveCallback) (v interface{}
 		r := c.buffer[0]
 		c.buffer[0] = nil
 		c.buffer = c.buffer[1:]
+
+		// Move blocked sends into buffer
+		if len(c.blockedSends) > 0 {
+			b := c.blockedSends[0]
+			c.blockedSends[0] = nil
+			c.blockedSends = c.blockedSends[1:]
```
Comment (yycptt) [inline internal/internal_workflow.go]: Create a popBlockedSends function for these three lines?

File: internal/internal_workflow.go [hunk 2]
```diff
@@ -584,6 +585,17 @@ func (c *channelImpl) receiveAsyncImpl(callback *receiveCallback) (v interface{}
 		r := c.buffer[0]
 		c.buffer[0] = nil
 		c.buffer = c.buffer[1:]
+
+		// Move blocked sends into buffer
+		if len(c.blockedSends) > 0 {
+			b := c.blockedSends[0]
+			c.blockedSends[0] = nil
```
Comment (andrewjdawson2016) [inline internal/internal_workflow.go]: Does this line do anything given the next line?

File: internal/internal_workflow_test.go [hunk 3]
```diff
@@ -757,6 +758,32 @@ func (s *WorkflowUnitTest) Test_CorruptedSignalOnClosedChannelWorkflow_Receive_S
 	s.EqualValues(0, len(result))
 }
 
+func bufferedChanWorkflowTest(ctx Context) error {
```
Comment (andrewjdawson2016) [inline internal/internal_workflow_test.go]: I know this was the customer's repro but there are two things I find odd about this repro
1. Calling receive inside the for loop that is doing the sending is odd. 
2. I do not think we need the // Refill the buffer loop in order to repro this. 

Can we use a simpler repro here so its clear that the bug occurs more generally and not just in this case.

Comment (andrewjdawson2016) [inline internal/internal_workflow_test.go]: Can we add a parameter to this workflow that includes buffer size? That way you can easily write another test for buffer size greater than one.

File: internal/internal_workflow_test.go [hunk 4]
```diff
@@ -757,6 +758,32 @@ func (s *WorkflowUnitTest) Test_CorruptedSignalOnClosedChannelWorkflow_Receive_S
...
+}
+
+func (s *WorkflowUnitTest) Test_BufferedChanWorkflow() {
+	env := s.NewTestWorkflowEnvironment()
+
+	env.ExecuteWorkflow(bufferedChanWorkflowTest)
+	s.True(env.IsWorkflowCompleted())
+	s.NoError(env.GetWorkflowError())
+}
+
```
Comment (andrewjdawson2016) [inline internal/internal_workflow_test.go]: Can we add a test or maybe even in the same test that passes buffer size > 1

File: internal/internal_workflow_test.go (hunk 3 above)
Comment (yycptt) [reply internal/internal_workflow_test.go]: Yeah I can change the bufferSize to a parameter.

File: internal/internal_workflow_test.go (hunk 4 above)
Comment (yycptt) [reply internal/internal_workflow_test.go]: Once bufferSize is a parameter, I can use a loop to test more cases.

File: internal/internal_workflow.go [hunk 5]
```diff
@@ -584,6 +585,17 @@ func (c *channelImpl) receiveAsyncImpl(callback *receiveCallback) (v interface{}
...
+			c.blockedSends = c.blockedSends[1:]
+			if b.fn() {
```
Comment (mfateev) [inline internal/internal_workflow.go]: I think it should iterate through blockedSends until b.fn() returns true. In this case if it returns false nothing is added to the buffer.

Comment (mfateev) [reply internal/internal_workflow.go]: I think you have to use Select that matches a different branch than send to reproduce this edge case.

File: internal/internal_workflow.go (hunk 2 above)
Comment (yycptt) [reply internal/internal_workflow.go]: No actually. I am just following existing pattern in the code. See Line 606.

File: internal/internal_workflow_test.go (hunk 3 above)
Comment (yycptt) [reply internal/internal_workflow_test.go]: Sure. I can create a simple one.

File: internal/internal_workflow.go (hunk 2 above)
Comment (mfateev) [reply internal/internal_workflow.go]: I believe it would retain memory without this assignment.

File: internal/internal_workflow.go (hunk 5 above)
Comment (yycptt) [reply internal/internal_workflow.go]: Updated and created a test case for it.

**Pull request merged**